### ISSUE TITLE
Feedback wanted

### ISSUE THREAD

**Issue by djetelina** [iScrE4m/pyArtifact#1]
I'll soon update the readme to call this an alpha state. With docs written and the Valve's API coverage being 100%, I'll be just fixing small stuff and writing tests, until I know in which direction I should be moving. Not writing any application with it myself, it's hard to determine what needs work for users and what's fine as is.

If you have any feedback to offer, even if it's just your feelings/opinions, I'd be very glad to hear them - feel free to comment on this issue with anything. If you have a more specific request, don't be scared to create a new issue.

**Comment by max3a3** [comment]
I was testing some card, the creep card 'Satyr Magician' id 10137, has active ability card which have the same name, so I can not access the creep card by name

**Comment by djetelina** [comment]
@max3a3 Thanks a lot! I'll think about how to deal with those cases. Internally, it means I can't use flat dict to store them and quickly look up by names, there are a few solutions for that, but then I would like not to complicate the API too much, if you got a list of the two with get, it'd mean every user would have to have a logic to handle those special cases... I'll think about it :)

**Comment by leileix2** [comment]
Thanks for making this, it's really useful! I'm trying it by making a reddit bot.

On that note, I noticed some decks doesn't return a 'name' to me, only an empty string. Is it an issue with decoding or? Admittedly I'm a newbie at this so I can't figure it out. Some sample decks:

- ADCFYIjfWG8AYoYEDhdQirdATYDAUUCDaoCSEGHA0oQAkWRhC8B - Should be "Red Deck"
- ADCJW8AJX05uwFCKgGAuF1CI90BwwQGBg+MSgVyARIQERE7AVEKEFkYBw__ - Should be "Green / Black Deck"
- ADCJT8AJ7kCAIYZS7hd7t0BBAEKHkEBLgEJVgoBASIBEhxJD0YWBUssAQc_ - Should be "Black / Red / Blue Deck"

This works:
- ADCJWkTZX05uwGDCRV4XQGy3QGLmqUBg4GQJgGLGgO7AaABR3JlZW4vQmxhY2sgRXhhbXBsZQ__ - "Green/Black Example"
- ADCJQUQI30zuwEYg2ABeF1Bu94BmWIBTEkLtAKlAZakAYmHh0JsdWUvUmVkIEV4YW1wbGU_ - "Blue/Red Example"

I tried rewriting the decoding part by copying your code / referencing Valve's PHP one but I end up getting garbage text lmao. Thanks for looking into this :+1:

**Comment by djetelina** [comment]
@leileix2 I looked into the Green / Black Deck and it seems that it's a valid version 2 deck - so it's capable of  holding the name, it just isn't filled there. The code with the name would be:

`ADCJW8SJX05uwFCKgGAuF1CI90BwwQGBg+MSgVyARIQERE7AVEKEFkYB0dyZWVuIC8gQmxhY2sgRGVjaw__`

vs your original: 

`ADCJW8AJX05uwFCKgGAuF1CI90BwwQGBg+MSgVyARIQERE7AVEKEFkYBw__`

Even from the code you can see how it's different. ADCJW8 is a common start and that's okay, then there's a difference in A vs S. That is +- where the byte (3rd byte) for string length lays - in the original, it's 0. Then the contents are the same, until the end, where the name lays.

tl;dr: I don't know what's the origin of your deck codes, but they simply seem to include empty names.

**Comment by leileix2** [comment]
Thanks for the quick response @iScrE4m! I got the codes from various reddit comments specifically from pro decks used in the recently streamed tournament if I remember correctly. So maybe the deck codes were from an old version of Artifact, i.e. when Valve hasn't released an API yet.

Except for the "Red Deck". I generated that code using [Artibuff's deck builder](https://www.artibuff.com/decks/ADCFYIjfWG8AYoYEDhdQirdATYDAUUCDaoCSEGHA0oQAkWRhC8B/edit) today. [PlayArtifact shows the name though](https://playartifact.com/d/ADCFYIjfWG8AYoYEDhdQirdATYDAUUCDaoCSEGHA0oQAkWRhC8B) but like I speculated it's probably because Valve has an old (fallback?) way of decoding the deck, idk.

It looks like newer codes generated from the API will show the name anyways so this is definitely a non-issue. I just thought it was interesting to bring up haha!

Thanks again for creating this wrapper!

**Comment by djetelina** [comment]
@leileix2 The red deck is interesting, it is a version 1 deck code (version 2 of the same would be `ADCJYIAI31hvAGKGBA4XUIq3QE2AwFFAg2qAkhBhwNKEAJFkYQvAQ__`), so I suspect Valve might be analyzing the contents and showing Red Deck based on the colors present? It certainly can't hold the name in the deck string. Good thing to know though, if I'm ever bored I might try and implement the same thing - but it's not very high priority for me :)

**Comment by djetelina** [comment]
@max3a3 I've come up with a fix, please let me know if you think it's a good design, before I release it.

[Docs](https://pyartifact.readthedocs.io/en/latest/api_cards.html#pyartifact.api_sync.Cards.get)

[Test](https://github.com/iScrE4m/pyArtifact/commit/ba004b699682e06829ed08d726a19695081925b9#diff-c50fa84e02e59189db6a0952e5950194R32)

edit: For the curious mind, here's a list of affected cards - bigger than I thought!

- Sister of the Veil
- Rebel Decoy
- Mercenary Exiles
- Emissary of the Quorum
- Ravenhook
- Ravenous Mass
- Assassin’s Apprentice
- Satyr Magician
- Unsupervised Artillery
- Revtel Investments
- Aghanim’s Sanctum
- Escape Route
- Messenger Rookery
- Keenfolk Turret
- Steam Cannon
- Assassin’s Veil
- Phase Boots
- Blink Dagger
- Demagicking Maul
- Rumusque Vestments
- Keenfolk Musket
- Bracers of Sacrifice
- Helm of the Dominator
- Wingfall Hammer
- Book of the Dead
- Shiva’s Guard
- Horn of the Alpha
- Nyctasha’s Guard
- Apotheosis Blade

**Comment by max3a3** [comment]
The change to get() works for me, Thanks

**Comment by djetelina** [comment]
@max3a3 Glad to hear, sorry for the long wait, it's now up on pypi under the version 0.3.2

**Comment by kevinchua6** [comment]
Hey, im trying to use your pyArtifact decoder but I'm having some trouble.

When i use your decode_deck_string() function i get a list of ids for the cards, but is there a way to convert those ids into the card names? Or is there a way to convert the deck code directly into the card names?

I'm sorry if this is a stupid question, I'm really new at this.

`>>> from pyartifact import decode_deck_string
>>> deck_contents = decode_deck_string('ADCJQUQI30zuwEYg2ABeF1Bu94BmWIBTEkLtAKlAZakAYmHh0JsdWUvUmVkIEV4YW1wbGU_')
>>> print(deck_contents['name'])
Blue/Red Example
>>> print(deck_contents['heroes'])
[{'id': 4003, 'turn': 1}, {'id': 10006, 'turn': 1}, {'id': 10030, 'turn': 1}, {'id': 10033, 'turn': 3}, {'id': 10065, 'turn': 2}]

**Comment by djetelina** [comment]
Hello, there is. Using the [Deck.from_code()](https://pyartifact.readthedocs.io/en/stable/api_deck.html#pyartifact.deck.Deck.from_code), you'll be able to get access to the cards instanced with all of their information as python objects. Short example could be:

```python
>>> from pyartifact import Cards, Deck
>>> Cards().load_all_sets()
>>> deck_contents = Deck.from_code('ADCJQUQI30zuwEYg2ABeF1Bu94BmWIBTEkLtAKlAZakAYmHh0JsdWUvUmVkIEV4YW1wbGU_')
>>> names = set([c.name for c in deck_contents.expand_cards()])
>>> print(f'Deck includes: {", ".join(names)}')
Deck includes: Ventriloquy, Time of Triumph, Blade of the Vigil, Viscous Nasal Goo, Phase Boots, Foresight, Traveler's Cloak, Red Mist Pillager, Fighting Instinct, Echo Slam, Eclipse, Thundergod's Wrath, Cunning Plan, Red Mist Maul, Routed, New Orders, Clear the Deck, Conflagration, Leather Armor, Tower Barrage
```

**Comment by kevinchua6** [comment]
Thanks so much!
